### ISSUE TITLE
Export report

### ISSUE THREAD

**Issue by jphorn** [roytanck/rt-plugin-report#15]
It would be nice of the generated table could be exported as PDF, CSV, Google Sheets, etc. This would help in sending a report of installed (and activated) plugins to fellow plugin authors to better support in diagnosing what plugins are running and which version.

**Comment by roytanck** [comment]
This was added in version 1.4 of the plugin which is available through wordpress.org.

**Comment by jphorn** [comment]
Unfortunately this seems broken for me (macOS/Safari and macOS/Chrome). In both browsers I'm unable to download the XLS file when using the export button. Clicking the button doesn't initiate any downloads and no alert is being shows. So essentially it does nothing at the moment.

**Comment by roytanck** [comment]
Could it be that you're running a plugin with special (non-latin) characters in either the name or the author's name? If so, please let me know which plugin, so I can attempt to recreate the problem.

(I'll try to get this fixed asap, but I'm currently not able to spend a lot of time on side projects like this)

**Comment by jphorn** [comment]
I can send you a screenshot of my installed plugins privately.

**Comment by roytanck** [comment]
Please do. My email is roy@tanck.nl .

**Comment by Zodiac1978** [comment]
I'm happy to help if I can replicate the issue. Have you found the culprit here? A specific plugin?

**Comment by roytanck** [comment]
The reporter did email me a screenshot as offered. This particular WordPress install had a considerable number of plugins installed. I saw no really exotic characters, just an occasional ö, ü and é.

That being said, character encoding appears to be a real problem with javascript's `btoa()` function.

I installed some of the same plugins with these characters ("Mashshare" and "AddQuicktag") on my dev install, but the table exported fine for me. Please feel free to see if you can reproduce this issue on your end. Perhaps a dummy plugin with a ß in the name/author will do the trick?

**Comment by Zodiac1978** [comment]
I can reproduce the issue in Chrome (but not in Firefox). Researching now.

**Comment by roytanck** [comment]
@jphorn Woudl the exported file still be useful to you if the red/green colors and the hyperlinks were missing? It's quite easy to do a reliable CSV export from javascript, which could imported into any spreadsheet program, but this would lack all formatting.

**Comment by jphorn** [comment]
Hmmm. Hyperlinks would be nice to have. Can you export them separately  in another column? Active/inactive is still exported?

**Comment by roytanck** [comment]
Yes, activation info is still exported. Essentially, you get the text that's in each cell, stripped of any HTML or CSS.

I'll see about adding the link if it's available. With a bit of trickery, that might be achievable.

For now I'm just assessing this as easier to maintain alternative to adding a 3rd party library to the plugin.

**Comment by roytanck** [comment]
This is an example of the output of the custom-built CSV export I whipped up. (Had to zip it because GitHub does not accept CSV uploads).

[plugin-report-2020-02-27.zip](https://github.com/roytanck/rt-plugin-report/files/4394197/plugin-report-2020-02-27.zip)

**Comment by jphorn** [comment]
That seems weird to me. Last line is all garbled and Activated column results show as `0/3` or `2/3`. Is that correct?

Also, could you please add column headers for `plugin_url` and `plugin_author_url`?

<details>
<summary>Show the table export</summary>
Name;;Author;;Activated;Installed version;Last update;Tested up to WP version;Rating;
AddQuicktag;http://bueltge.de/wp-addquicktags-de-plugin/120/;Frank Bültge;https://bueltge.de;0/3;2.5.3;3 weeks;;96%;
Classic Editor;https://wordpress.org/plugins/classic-editor/;WordPress Contributors;https://github.com/WordPress/classic-editor/;0/3;1.5;4 months;5.3.2;100%;
Disable Comments;https://wordpress.org/plugins/disable-comments/;Samir Shah;http://www.rayofsolaris.net/;2/3;1.10.2;8 months;5.3.2;98%;
kk Star Ratings;https://github.com/kamalkhan/kk-star-ratings;Kamal Khan;http://bhittani.com;0/3;4.1.3;3 months;5.2.5;80%;
Koko Analytics;https://www.kokoanalytics.com/;ibericode;https://ibericode.com/;Network activated;1.0.9 (1.0.10 available);4 days;5.4;100%;
Mail Caesar!;;Roy Tanck;https://roytanck.com/;1/3;1.1;2 weeks;5.3.2;No data available;
Mashshare Share Buttons;https://www.mashshare.net;René Hermenau;https://www.mashshare.net;0/3;3.7.3;7 days;5.4;96%;
Plugin Report;https://roytanck.com/;Roy Tanck;https://roytanck.com;Network activated;1.6.1;5 days;5.4;100%;
Regenerate Thumbnails;https://alex.blog/wordpress-plugins/regenerate-thumbnails/;Alex Mills (Viper007Bond);https://alex.blog/;Network activated;3.1.2 (3.1.3 available);3 weeks;5.3.2;92%;
RT Filter Page List;http://www.roytanck.com;Roy Tanck;http://www.roytanck.com;0/3;0.9.4;2 weeks;5.4;100%;
RT List Blogs;;Roy Tanck;;0/3;0.9;No data available;No data available;No data available;
Tag Pages;http://burobjorn.nl;Bjorn Wijers ;http://burobjorn.nl;1/3;1.0.2;2 months;5.3.2;98%;
WP Blogmesh;http://www.roytanck.com;Roy Tanck;;0/3;0.5 (0.8 available);2 weeks;5.4;No data available;
Lássuk: működik-e: ez:;;ဒီဟာအလုပ်လုပ်လားကြည့်ရအောင်;;讓我們看看這是否有效;بیایید ببینیم که آیا این کار می کند;Бул иштей тургандыгын көрөлү;இது செயல்படுகிறதா என்று பார்ப்போம்;これがうまくいくか見てみましょう;
</details>

**Comment by roytanck** [comment]
Oh, sorry. Forgot to mention that there's an extra row with test data in foreign languages (Hindi, Japanese, etc) to test if non-latin characters are preserved.

The "2/3" means the plugin in active on two out of a total of three sites within a multisite network. Single site installs will simply show activated/not activated.

**Comment by jphorn** [comment]
OK. Makes sense. Looks good to me then :-)

**Comment by roytanck** [comment]
Thank you. I'll see a about adding those column headers if we go with CSV.

**Comment by Zodiac1978** [comment]
@roytanck If you go the CSV route – I would recommend this, because I don't like proprietary software) – I would recommend using this trick to add the header info:

In [RFC4180](https://tools.ietf.org/html/rfc4180) there is an interesting line stating:

> ...the presence or absence of the header line should be indicated via the optional "header" parameter of this MIME type...

The correct usage is explained here:
https://stackoverflow.com/questions/15457752/proper-syntax-for-optional-header-parameter-for-text-csv-mimetype

With this info we can assure the column headers are used in Office software. After reading the spec this seems the only generic possibility to add a feature to a CSV.

**Comment by roytanck** [comment]
@Zodiac1978 I added another feature branch that replaces the XLS export with a homebrew CSV one. Initially I chose XLS because it appeared to preserve styling. This worked in GDrive, but not in LibreOffice. In general, the "html table wrapped in Excel headers" export seemed to be hit-and-miss between browsers and spreadsheet programs. I agree that CSV is a better option.

I added the "header" header to the MIME type, but this does not add a real header row in GDrive/LibreOffice for me.

Obviously, feel free to test this.

**Comment by Zodiac1978** [comment]
I've checked the XLS files on a real Microsoft Excel (Mac version) and got this warning:

> Warning
> The file format and the file extension of 'plugin-report-2020-02-27-x.xls' do not match. The file may be damaged or not secure. You should not open it if you do not trust its source. Do you still want to open the file?

After clicking "OK" the file opens correctly and shows up all data correctly and with styles/colors. Just FYI.

**Comment by roytanck** [comment]
That's weird, since the extension is modified here:

https://github.com/roytanck/rt-plugin-report/commit/15f3f0bd661f663e57732180ae3067498b65f136#diff-427fae7125014b912a15fa857ea236a2R111

You should be getting a csv file. Perhaps the old js file is used from browser cache?

**Comment by Zodiac1978** [comment]
No ... I'm just answering to a different question first:
> If you would be able to test this in Numbers (and possibly Excel itself) that would be really helpful.
from #33

**Comment by roytanck** [comment]
Sorry, I was confused. That warning is to be expected, because the XLS file does not actually contain Excel data. It needs to convert the embedded chunk of HTML data to its internal format.

All the simple javascript approaches I've seen do this, and other programs will import without the warning. Still, it remains a somewhat dirty trick.

**Comment by Zodiac1978** [comment]
Yes, I understand :)
Just wanted to report back what I saw after testing.

Now to the CSV file:
Works great and flawless so far. Everyone understands CSV. Encoding is fine for me (üöäß showing correctly). Formatting is of course not available.

LibreOffice is working fine, but no header.
Numbers is using the first line as a header. Not sure if this is because of the header MIME type or the default for CSV in Numbers. 

One thing: The export button is showing immediately. I would recommend to hide the button until the data is complete. Just move line 49-56 directly below line 23, like I did in the PR for the XLS export fix.
